ddl/partition.go

```diff
@@ -978,6 +978,44 @@ func onTruncateTablePartition(d *ddlCtx, t *meta.Meta, job *model.Job) (int64, e
 		}
 	}
 
+	var rules []*placement.RuleOp
+
+	rules, err = infosync.GetPlacementRules(nil)
+	if err != nil {
+		job.State = model.JobStateCancelled
+		return ver, errors.Wrapf(err, "failed to retrieve placement rules from PD")
+	}
+
+	// TODO: maybe add a middle state
+	newRules := make([]*placement.RuleOp, 0, len(oldIDs))
+	for i, oldID := range oldIDs {
+		prefix := fmt.Sprintf("%d_t%d_p%d", job.SchemaID, tblInfo.ID, oldID)
+		for _, rule := range rules {
+			if strings.HasPrefix(rule.ID, prefix) {
+				// delete the old rule
+				newRules = append(newRules, &placement.RuleOp{
+					Action: placement.RuleOpDel,
+					Rule: &placement.Rule{
+						GroupID: placement.RuleDefaultGroupID,
+						ID:      rule.ID,
+					},
+				})
+
+				// add the new rule
+				rule.Action = placement.RuleOpAdd
+				rule.ID = fmt.Sprintf("%d_t%d_p%d_%s_%d_%d", job.SchemaID, tblInfo.ID, newPartitions[i].ID, rule.Role, job.ID, i)
+				newRules = append(newRules, rule)
+				break
+			}
+		}
+	}
+
+	err = infosync.UpdatePlacementRules(nil, newRules)
```

Comment:
What if the DDL worker is down before updating placement rules?
After rebooting, although the worker continues to drop partitions, it doesn't know what rules to update.

Reply:
Pardon, I did not get it. Table meta is not updated if it is failed. It will retry to get all rules from PD, which is not updated.

Reply:
> which is not updated
Why? The rules are updated after `UpdatePlacementRules`.

Reply:
`UpdatePlacementRules` is transactional. It will cancel failed operations.